### PR TITLE
API: Lay groundwork & Add Hackathons resource

### PR DESCRIPTION
**Pull request by garyhtou**
Introducing the ✨Hackathons API✨!!

The API has the following features:
- **Versioned.** Thanks to https://github.com/bwillis/versioncake, it's simple to add a new version in a DRY manner
- **Well-defined shapes.** All objects, paginated responses, and error responses have a consistent body
	- All objects have an: `id` (hashid), `type` (model), `created_at`, `link.self` (URL to API resource)
	- All paginated responses have a: `data` array, `meta.total_count` (total records), `meta.total_pages`, `links.next`, `links.prev`, `links.first`, `links.last`, `links.self`
	- All error responses have a: `type`, `title`, `detail`, `status` (HTTP code)
		- In development, the response also contains a `development_debug` key which has the error's class, inspect, and backtrace
- **Pagination.** With page-based pagination, the API returns a `links.next` key which contains a URL to the next page
- **Error Handling.** All errors are handled and respond as JSON (including unexpected internal server errors)
- Also roughly follows [JSON:API](https://jsonapi.org/)https://jsonapi.org/

## Example responses

<details>
<summary><code>/api/v1/hackathons?per_page=1</code></summary>

Default page size is 100, max is 250
```json
{
   "data":[
      {
         "id":"dNuEgaE3N",
         "type":"hackathon",
         "name":"Assemble",
         "starts_at":"2023-08-05T00:00:00.000Z",
         "ends_at":"2023-08-07T00:00:00.000Z",
         "website":"https://assemble.hackclub.com",
         "modality":"in_person",
         "logo":"http://localhost:3000/rails/active_storage/representations/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBCTVdKandzPSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--1e9c5db09381e72593617b9c99ad63b423ab3f87/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaDdCem9MWm05eWJXRjBTU0lJY0c1bkJqb0dSVlE2RkhKbGMybDZaVjkwYjE5c2FXMXBkRnNIYVFHQWFRR0EiLCJleHAiOm51bGwsInB1ciI6InZhcmlhdGlvbiJ9fQ==--d0d5199a2231debcbddb3d87041296b55b32e5be/assemble_logo.png",
         "banner":"http://localhost:3000/rails/active_storage/representations/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBCRVN6Q0E0PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--bcec1d67c04615e8cef5f6047d44ed77782b35f5/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaDdCem9MWm05eWJXRjBTU0lJY0c1bkJqb0dSVlE2RkhKbGMybDZaVjkwYjE5c2FXMXBkRnNIYVFLQUIya0NPQVE9IiwiZXhwIjpudWxsLCJwdXIiOiJ2YXJpYXRpb24ifX0=--a181ba1964000bcdc69f0c88884de1dcbc9450ef/assemble.png",
         "location":{
            "city":null,
            "state":null,
            "country":null,
            "longitude":null,
            "latitude":null
         },
         "created_at":"2023-07-20T00:04:38.189Z",
         "links":{
            "self":"http://localhost:3000/api/v1/hackathons/dNuEgaE3N"
         }
      }
   ],
   "links":{
      "first":"http://localhost:3000/api/v1/hackathons?per_page=1&page=1",
      "prev":null,
      "next":"http://localhost:3000/api/v1/hackathons?per_page=1&page=2",
      "last":"http://localhost:3000/api/v1/hackathons?per_page=1&page=12",
      "self":"http://localhost:3000/api/v1/hackathons?per_page=1&page=1"
   },
   "meta":{
      "total_count":12,
      "total_pages":12
   }
}
```
</details>

<details>
<summary><code>/api/v1/hackathons/dNuEgaE3N</code></summary>

```json
{
  "id": "dNuEgaE3N",
  "type": "hackathon",
  "name": "Assemble",
  "starts_at": "2023-08-05T00:00:00.000Z",
  "ends_at": "2023-08-07T00:00:00.000Z",
  "website": "https://assemble.hackclub.com",
  "modality": "in_person",
  "logo": "http://localhost:3000/rails/active_storage/representations/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBCTVdKandzPSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--1e9c5db09381e72593617b9c99ad63b423ab3f87/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaDdCem9MWm05eWJXRjBTU0lJY0c1bkJqb0dSVlE2RkhKbGMybDZaVjkwYjE5c2FXMXBkRnNIYVFHQWFRR0EiLCJleHAiOm51bGwsInB1ciI6InZhcmlhdGlvbiJ9fQ==--d0d5199a2231debcbddb3d87041296b55b32e5be/assemble_logo.png",
  "banner": "http://localhost:3000/rails/active_storage/representations/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBCRVN6Q0E0PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--bcec1d67c04615e8cef5f6047d44ed77782b35f5/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaDdCem9MWm05eWJXRjBTU0lJY0c1bkJqb0dSVlE2RkhKbGMybDZaVjkwYjE5c2FXMXBkRnNIYVFLQUIya0NPQVE9IiwiZXhwIjpudWxsLCJwdXIiOiJ2YXJpYXRpb24ifX0=--a181ba1964000bcdc69f0c88884de1dcbc9450ef/assemble.png",
  "location": {
    "city": null,
    "state": null,
    "country": null,
    "longitude": null,
    "latitude": null
  },
  "created_at": "2023-07-20T00:04:38.189Z",
  "links": {
    "self": "http://localhost:3000/api/v1/hackathons/dNuEgaE3N"
  }
}
```
</details>

<details>
<summary><code>Error Response (Object not found)</code></summary>

```json
{
  "type": "not_found_error",
  "title": "Object not found",
  "detail": "The requested Hackathon with id='dNuEgaE3' could not be found.",
  "status": 404
}
```
</details>

<details>
<summary><code>Error Response (Internal server error)</code></summary>

```json
{
  "type": "internal_server_error",
  "title": "An error has occurred",
  "detail": "Please try again later. If the problem persists, please contact bank@hackclub.com.",
  "status": 500
}
```
</details>

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -25,7 +25,12 @@ gem "aws-sdk-s3", require: false
 gem "image_processing", ">= 1.2"
 gem "active_storage_validations"
 
+# API
 gem "jbuilder" # JSON templating
+gem "versioncake"
+gem "jbuilder_pagination_plus", require: "jbuilder/pagination" # Depends on Kaminari and Jbuilder
+
+gem "kaminari"
```
Comment (northeastprince) [inline Gemfile:33]: Thoughts on using [pagy](https://github.com/ddnexus/pagy) instead? Aside from the reasons on the README, it'll also negate the need for `jbuilder_pagination_plus`, which doesn't look widely used or maintained anyway.

File: app/controllers/api/concerns/errors.rb [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    # Handle any custom API errors. Generally when creating and raising these
+    # errors, we add additional context such as the `title` and `detail` as a
+    # part of instantiating the error.
+    rescue_from Api::Error, with: :api_error
+
+    # Handle any non-custom API errors. These require a custom handler to add
+    # more context such as the `title` and `detail`.
+    rescue_from VersionCake::UnsupportedVersionError, with: :unsupported_version
+    rescue_from ActiveRecord::RecordNotFound, with: :not_found
```
Comment (northeastprince) [inline app/controllers/api/concerns/errors.rb:17]: This already happens?

File: app/controllers/api/concerns/errors.rb [hunk 3]
```diff
@@ -0,0 +1,77 @@
+module Api::Concerns::Errors
```
Comment (northeastprince) [inline app/controllers/api/concerns/errors.rb:1]: All this fanfare around overriding Rails' error handling seems kinda redundant, we're basically just restating the status code and some other obvious information throughout 5 additional files.

File: app/helpers/api_helper.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    end
+    json.meta do
+      json.count collection.total_count
+    end
+  end
+
+  # Shape for an object
+  def obj(json, object)
+    json.id object.hashid
+    json.type object.class.try(:api_type) || object.class.name.underscore
```
Comment (northeastprince) [inline app/helpers/api_helper.rb:18]: The route should always describe what should be returned.

File: app/helpers/api_helper.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  end
+
+  # API URL for an object. By default, the api_version of the generated URL is
+  # the same as the current request's version.
+  def api_url_for(object, **options)
+    polymorphic_url([:api, object], api_version: @request_version, **options)
+  end
+
+  # Permanent URL for an attached file or variant
+  def file_url_for(attachable, variant = nil)
```
Comment (northeastprince) [inline app/helpers/api_helper.rb:35]: ActiveStorage provides link generators iirc

File: Gemfile (hunk 1 above)
Comment (garyhtou) [reply Gemfile]: I remember we had tried it for https://github.com/hackclub/bank/ but found no performance improvement. Nevertheless, let's give it a shot!

[`611f0b3` (#57)](https://github.com/hackclub/hackathons-backend/pull/57/commits/611f0b304d83b6e15c06f72d481a9008d369a5d7)

File: app/controllers/api/concerns/errors.rb (hunk 3 above)
Comment (garyhtou) [reply app/controllers/api/concerns/errors.rb:1]: Notes from meeting:
- We chatted about how Rails doesn't provide the best error handling for JSON responses
- This code handles serializing error objects into JSON responses, which is something Rails doesn't provide when we add additional fields such as `title` and `detail`.
- For more information regarding JSON error responses, see https://jsonapi.org/format/#errors

File: app/helpers/api_helper.rb (hunk 4 above)
Comment (garyhtou) [reply app/helpers/api_helper.rb:32]: Notes from meeting:

To address the previous comment, `type`s are helpful in the case of nested objects. For example,
```json
{
  "id": "dNuEgaE3N",
  "type": "hackathon",
  "name": "Assemble",
  "subscribers": [
    {
      "id": "sl3jsdfLs",
      "type": "user",
      "name": "Gary Tou"
    }
  ]
}
```

Using class names for `type` strongly couples the API type to the backend implementation. Ideally, the API is decoupled from the Rails model implementation. This way, if we need to refactor the models, the API can still provide consistent types without breaking changes. Also, as with many fields intended for programmatic use, they are lowercase (although there isn't an exact standard around this).

We had discussed three possible implementations:
- `object.class.name.underscore`
	- `Hackathon` -> `hackathon`
	- `Hackathon::Digest` -> `hackathon/digest`
	- `Hackathon::ExampleClass` -> `hackathon/example_class`
	- Pros
		- Preserves namespace vs class name (also a con)
	- Cons
		- Preserving namespace vs class somewhat couples implementation
		- The slash is kinda weird/ugly
- `object.class.name.underscore.parameterize`
	- `Hackathon` -> `hackathon`
	- `Hackathon::Digest` -> `hackathon-digest`
	- `Hackathon::ExampleClass` -> `hackathon-example_class`
	- Pros
		- This is how gems are named! `-` are modules and `_` separate pascal case
	- Cons
		- Preserving namespace vs class somewhat couples implementation
- `object.class.name.underscore.gsub("/", "_")`
	- `Hackathon` -> `hackathon`
	- `Hackathon::Digest` -> `hackathon_digest`
	- `Hackathon::ExampleClass` -> `hackathon_example_class`
	- Pros
		- It's clean lookin' 🤷‍♂️
	- Cons
		- Chance of unintentional `type` conflict. For example, `Hackathon::ExampleClass` and `HackathonExample::Class` have the same `type`.

**We'll be going with `object.class.name.underscore.parameterize`**

Comment (garyhtou) [reply app/helpers/api_helper.rb]: [`c8458ed` (#57)](https://github.com/hackclub/hackathons-backend/pull/57/commits/c8458ed8a52271c56aa7c6408511c2cf3fb2b261)

File: app/helpers/api_helper.rb (hunk 5 above)
Comment (garyhtou) [reply app/helpers/api_helper.rb:49]: ### [`ActionView::Helpers.url_for`](https://apidock.com/rails/ActionView/Helpers/UrlHelper/url_for)
```ruby
url_for hackathon.logo
```
Doesn't provide the full URL. By default, it's `only_path: true`. However, I can't seem to properly pass in a `host` to get the full URL. Note that this implementation is actually different than `Rails.application.routes.url_helpers.url_for` from `ActionDispatch::Routing::RouteSet`

### [`ActionDispatch::Routing::RouteSet.url_for`](https://api.rubyonrails.org/classes/ActionDispatch/Routing/UrlFor.html)

```ruby
Rails.application.routes.url_helpers.url_for(hackathon.logo)
```

This is what we want! It's kinda of long tho

### `polymorphic_url`

```ruby
polymorphic_url(hackathon.logo)
```

This also works as expected! Full path and everything

---

So as a result, either `Rails.application.routes.url_helpers.url_for` or `polymorphic_url` works. Inside `file_url_for` I had provided a param to make variants cleaner.

```ruby
# before
file_url_for hackathon.logo.variant(:small)

# after
file_url_for hackathon.logo, :small
```

**Pull request merged**